library/cgi/escapeURIComponent_spec.rb

```diff
@@ -0,0 +1,41 @@
+require_relative '../../spec_helper'
+require 'cgi'
+
+describe "CGI.escapeURIComponent" do
+  ruby_version_is "3.2" do
+    it "escapes whitespace" do
+      str = "&<>\" \xE3\x82\x86\xE3\x82\x93\xE3\x82\x86\xE3\x82\x93"
+      CGI.escapeURIComponent(str).should == '%26%3C%3E%22%20%E3%82%86%E3%82%93%E3%82%86%E3%82%93'
+    end
+
+    it "does not escape with unreserved characters" do
+      str = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~"
+      CGI.escapeURIComponent(str).should == "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~"
+    end
+
+    it "with invalid bytesequence" do
+      str = "\xC0\<\<".force_encoding("UTF-8")
+      CGI.escapeURIComponent(str).should == "%C0%3C%3C"
+    end
+
+    it "nil raises a TypeError" do
+      -> { CGI.escapeURIComponent(nil) }.should raise_error(TypeError)
+    end
+
+    it "empty string" do
+      CGI.escapeURIComponent("").should == ""
+    end
+
+    it "single whitespace" do
+      CGI.escapeURIComponent(" ").should == "%20"
+    end
+
+    it "double whitespace" do
+      CGI.escapeURIComponent("  ").should == "%20%20"
+    end
+
+    it "preserves encoding" do
+      CGI.escapeURIComponent("whatever".force_encoding("ASCII-8BIT")).encoding.should == Encoding::ASCII_8BIT
+    end
+  end
+end
```

Comment:
**suggestion:** we could add some other cases:
- for implicit type conversion of an argument (to String with method `#to_str`)
- for ascii-compatible String and ascii-not-compatible (it treats a String as a bytes sequence and converts bytes, not characters)

Reply:
I'm a bit uncertain here. I added a spec for the implicit type conversion, but I'm not sure if I understood it correctly what you suggested. Please let me know if I misunderstood.

I could use a pointer regarding the ascii-not-compatible string idea... how could I approach this? Thanks for your help, I appreciate it!

Reply:
Regarding implicit conversion - you are correct. Actually you may look for similar test cases, e.g. here:

https://github.com/ruby/spec/blob/db5f026014a1153f03414668e27071676a0be2c2/core/basicobject/instance_eval_spec.rb#L287-L301

Reply:
>for ascii-compatible String and ascii-not-compatible (it treats a String as a bytes sequence and converts bytes, not characters)

> I could use a pointer regarding the ascii-not-compatible string idea... how could I approach this?

I meant that most of the test cases use ASCII characters. But the method actually handles multibyte characters, e.g. UTF-8/UTF-16/UTF-32 and just converts their bytes one by one (ignoring "characters" that could contain several bytes):

```ruby
CGI.escapeURIComponent("β")
# => "%CE%B2"

"β".bytes.map {|c| c.to_s(16) }
# => ["ce", "b2"]
```